assets/src/components/ladder.tsx

```diff
@@ -60,18 +64,77 @@ const Ladder = ({
       ? timepoints.slice().reverse()
       : timepoints
 
+  const ladderVehiclePositions: LadderVehiclePosition[] = putIntoLanes(
+    vehicles.map(vehicle => {
+      const vehicleDirection: VehicleDirection = vehicleDirectionOnLadder(
+        vehicle,
+        ladderDirection
+      )
+      const x =
+        vehicleDirection === VehicleDirection.Up
+          ? 47
+          : -(48 + widthOfVehicleGroup)
+      const y =
+        timepointStatusY(
+          vehicle.timepoint_status,
+          timepoints,
+          timepointSpacingY,
+          vehicleDirection
+        ) || -10
+
+      return {
+        vehicle,
+        vehicleDirection,
```

Comment:
There's some interesting algorithm stuff, but it's pretty tied up in the vehicle stuff. Do you think it'd be possible to split apart? Maybe something like
`putIntoLanes = (input: [{y: number, payload: whatever}], minSpaceBetween: number): [{y: number, payload: whatever, lane: number}]`
Maybe you could partition the up and down vehicle and call it separately on the two sides so that the algorithmic stuff doesn't need to think about `VehicleDirection` at all? I think it'd end up being easier to reason about and test if the lane algorithm didn't need to think about vehicle stuff, and if all the vehicle logic didn't have any lane algorithm mixed in.

Reply:
I think if I understand what you are describing it gets to why I wanted to pull out a view model in the first place, to get all the logic out of the components. I think the problem is that this function here doesn't belong here, it belongs in `ladderVehiclePosition`. What if we passed `vehicles` and `timepointStatusY` into a function in `ladderVehiclePosition` that is basically this function, but moving the `x` calculation to aftert we figure out the lanes? I can do that work to show it more concretely, but does that sound like it would be moving in a better direction?

Reply:
I think the view model helps a lot, and moving the `x` calculation to after we figure out the lanes would make that even better.

I think we could also separate the algorithmic part of `putIntoLanes` from the view model. Like, `putIntoLanes`, `overlappingPreviousVehicles`, `firstOpenLane`, and friends don't actually need any information about the vehicle, just its y, so could be pulled into a separate file.

There's a good abstraction boundary there that would make it easier to see the view model and how it's made and used without having to wade through any of the algorithmic stuff.

Reply:
Take a fresh look: I extracted all that logic out of the Ladder component and made the internal logic much more pure about what data it requires (with internal interfaces to demonstrate that).

Reply:
One small additional update: I curried some timepointStatusY variables to avoid passing them around unnecessarily.